runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/logging/JulHandlerPrintStreamAdapterFactory.java

```diff
@@ -37,114 +43,328 @@
-   */
-  private static class JulHandlerAdapterOutputStream extends OutputStream {
+  private static class JulHandlerPrintStream extends PrintStream {
     private static final String LOGGING_DISCLAIMER =
         String.format(
             "Please use a logger instead of System.out or System.err.%n"
                 + "Please switch to using org.slf4j.Logger.%n"
                 + "See: https://cloud.google.com/dataflow/pipelines/logging");
-    // This limits the number of bytes which we buffer in case we don't see a newline character.
-    private static final int BUFFER_LIMIT = 1 << 14; // 16384 bytes
-    private static final byte[] NEW_LINE = System.lineSeparator().getBytes(StandardCharsets.UTF_8);
+    // This limits the number of bytes which we buffer in case we don't have a flush.
+    private static final int BUFFER_LIMIT = 1 << 10; // 1024 chars
 
     /** Hold reference of named logger to check configured {@link Level}. */
-    private Logger logger;
+    private final Logger logger;
 
-    private Handler handler;
-    private String loggerName;
-    private ByteArrayOutputStream baos;
-    private Level messageLevel;
-    private int matched = 0;
+    private final Handler handler;
+    private final String loggerName;
+    private final StringBuilder buffer;
+    private final Level messageLevel;
+    private final CharsetDecoder decoder;
 
-    private JulHandlerAdapterOutputStream(Handler handler, String loggerName, Level logLevel) {
+
+    private JulHandlerPrintStream(Handler handler, String loggerName, Level logLevel) {
+      super(
+          new OutputStream() {
+            @Override
+            public void write(int i) throws IOException {
+              throw new RuntimeException("All methods should be overwritten so this is unused");
+            }
+          });
       this.handler = handler;
       this.loggerName = loggerName;
       this.messageLevel = logLevel;
       this.logger = Logger.getLogger(loggerName);
-      this.baos = new ByteArrayOutputStream(BUFFER_LIMIT);
+      this.buffer = new StringBuilder(BUFFER_LIMIT);
+      this.decoder = Charset.defaultCharset().newDecoder();
+    }
+
+    @Override
+    public void flush() {
+      publish(flushToString());
+    }
+
+    private synchronized String flushToString() {
+      if (buffer.length() > 0 && buffer.charAt(buffer.length() - 1) == '\n') {
+        buffer.setLength(buffer.length() - 1);
+      }
+      String result = buffer.toString();
+      buffer.setLength(0);
+      return result;
+    }
+
+    @Override
+    public void close() {
+      flush();
+    }
+
+    @Override
+    public boolean checkError() {
+      return false;
+    }
+
+    @Override
+    public synchronized void write(int i) {
+      buffer.append(i);
+    }
+
+    @Override
+    public void write(byte[] a, int start, int limit) {
+      CharBuffer decoded;
+      synchronized (decoder) {
+        try {
+          decoded = decoder.decode(ByteBuffer.wrap(a, start, limit));
```

Comment:
You need to use the stateful version of the CharsetDecoder.decode since this method assumes that the byte[] range represents a whole valid String.

In the other methods you would need to finish the decoding if there is a partial decoding in flight.

Reply:
Thanks for catching. I added a unit test for raw bytes and fixed to keep a buffer of remainders.